ext/objspace/objspace_dump.c

```diff
@@ -37,22 +40,104 @@ struct dump_config {
     unsigned int full_heap: 1;
 };
 
-PRINTF_ARGS(static void dump_append(struct dump_config *, const char *, ...), 2, 3);
 static void
-dump_append(struct dump_config *dc, const char *format, ...)
+dump_flush(struct dump_config *dc)
 {
-    va_list vl;
+    if (dc->buffer_len) {
+        if (dc->stream) {
+            fwrite(dc->buffer, 1, dc->buffer_len, dc->stream);
+        }
+        else if (dc->string) {
+            rb_str_cat(dc->string, dc->buffer, dc->buffer_len);
+        }
+        dc->buffer_len = 0;
+    }
+}
+
+static inline void
+buffer_inc_capa(struct dump_config *dc, unsigned long requested)
+{
+    unsigned long required;
+
+    if (requested + dc->buffer_len >= dc->buffer_capa) {
+        dump_flush(dc);
+    }
+
+    for (required = dc->buffer_capa; requested > required - dc->buffer_len; required <<= 1);
+
+    if (required > dc->buffer_capa) {
+        REALLOC_N(dc->buffer, char, required);
+        dc->buffer_capa = required;
+    }
+}
+
+static void buffer_append(struct dump_config *dc, const char *cstr, unsigned long len)
+{
+    if (len > 0) {
+        buffer_inc_capa(dc, len);
+        MEMCPY(dc->buffer + dc->buffer_len, cstr, char, len);
+        dc->buffer_len += len;
+    }
+}
+
+# define dump_append(dc, str) buffer_append(dc, (str), (long)strlen(str))
+
+PRINTF_ARGS(static void dump_appendf(struct dump_config *, const char *, ...), 2, 3);
+static void
+dump_appendf(struct dump_config *dc, const char *format, ...)
+{
+    va_list vl, vl2;
     va_start(vl, format);
+    va_copy(vl2, vl);
+
+    unsigned long capa_left = dc->buffer_capa - dc->buffer_len;
+    unsigned long required = vsnprintf(dc->buffer + dc->buffer_len, capa_left, format, vl);
 
-    if (dc->stream) {
-	vfprintf(dc->stream, format, vl);
+    if (required >= capa_left) {
+        buffer_inc_capa(dc, required + 1);
+        capa_left = dc->buffer_capa - dc->buffer_len;
+        vsnprintf(dc->buffer + dc->buffer_len, capa_left, format, vl2);
     }
-    else if (dc->string)
-	rb_str_vcatf(dc->string, format, vl);
+    dc->buffer_len += required;
 
     va_end(vl);
 }
 
+static void
+dump_appendc(struct dump_config *dc, char c)
+{
+    buffer_inc_capa(dc, 1);
+    dc->buffer[dc->buffer_len] = c;
+    dc->buffer_len++;
+}
+
+static const char hex_table[16] = {
+  '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'
+};
+
+static void
+dump_append_ref(struct dump_config *dc, VALUE ref)
+{
+    char buffer[16] = "\"0x";
+
+    buffer[ 3] = hex_table[(ref & 0xF00000000000) >> 44];
```

Comment:
This can cause longer-than-necessary output for 32bit environments.

Reply:
I suppose I can define two distinct implementations based on `SIZEOF_VALUE`.

Reply:
I pushed a new commits that fallbacks to `dump_appendf(dc, "\"0x%"PRIxVALUE"\"", ref);` if `SIZEOF_VALUE != SIZEOF_LONG_LONG`

It should be fairly easy to write the same optimization for 32bits systems, but I don't have one to test so I'd rather play it safe.

Reply:
Why are MSBs dropped?

Reply:
Because as far as I can tell that is how `"0x%"PRIxVALUE"` behaves, e.g on 2.7.1:

```ruby
>> puts JSON.parse(ObjectSpace.dump('foo'))['address']
0x7fd48c44cdf8
```

That's 12 hex digits so 48 bits not 64.

Reply:
It depends on systems.
Not portable and unacceptable.

Reply:
Ok, I'll use PRIxVALUE in all cases.

Reply:
Done. We're back to ~3x faster than baseline rather than 4x, though, but it's good enough already.